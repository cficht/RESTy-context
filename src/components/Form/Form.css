.Form {
  display: flex;
  justify-content: space-between;
  width: 100%;
  background-color: hotpink;

  form {
    display: flex;
    justify-content: space-evenly;
    padding: 10px;
    width: 100%;
  }

  .Left {
    display: flex;
    flex-direction: column;
    align-items: center;
    margin: 10px;
    padding: 10px;
    width: 40%;
    text-align: center;
    background-color: indianred;

    input[type=text] {
      margin: 10px;
      width: 80%;
    }

    input[type=radio] {
      margin: 0px 10px 0px 10px;
    }

    textarea {
      margin: 10px;
      width: 80%;
      height: 100px;
    }

    button {
      width: 20%;
    }
  }

  .Right {
    display: flex;
    flex-direction: column;
    align-items: center;
    margin: 10px;
    padding: 10px;
    width: 40%;
    background-color: indigo;
  }

  .AuthTop {
    width: 100%;
    text-align: center;
    background-color: khaki;
  }

  .AuthBot {
    width: 100%;
    padding: 50px;
    text-align: center;
    background-color: green;

    input{
      margin: 10px;
    }

    input[name=SET_TOKEN]{
      width: 100%;
    }
  }
}