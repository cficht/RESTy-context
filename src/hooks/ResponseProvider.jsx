import React, { createContext, useReducer, useContext } from 'react';
import PropTypes from 'prop-types';

const ResponseContext = createContext();

const initialState = {
  url: '',
  method: '',
  body: '',
  res: {},
  requests: []
};

export function reducer(state, action) {
  switch(action.type) {
    case 'SET_URL':
      return { ...state, url: action.payload };
    case 'SET_METHOD':
      return { ...state, method: action.payload };
    case 'SET_BODY':
      return { ...state, body: action.payload };
    case 'SET_RES':
      return { ...state, res: action.payload };
    case 'ADD_REQUESTS':
      return { ...state, requests: [...state.requests, action.payload] };
    default:
      return state;
  }
}

export const ResponseProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <ResponseContext.Provider value={{ state, dispatch }}>
      {children}
    </ResponseContext.Provider>
  );
};

ResponseProvider.propTypes = {
  children: PropTypes.node
};

export const useGlobalState = () => {
  const { state } = useContext(ResponseContext);
  return state;
};

export const useDispatch = () => {
  const { dispatch } = useContext(ResponseContext);
  return dispatch;
};

export const useUrl = () => {
  const { url } = useGlobalState();
  return url;
};

export const useMethod = () => {
  const { method } = useGlobalState();
  return method;
};

export const useBody = () => {
  const { body } = useGlobalState();
  return body;
};

export const useRes = () => {
  const { res } = useGlobalState();
  return res;
};

export const useRequests = () => {
  const { requests } = useGlobalState();
  return requests;
};
