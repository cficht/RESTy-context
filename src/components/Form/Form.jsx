import React from 'react';
import { useGlobalState, useDispatch } from '../../hooks/ResponseProvider';
import { fetchResponse } from '../../services/request';
import styles from './Form.css';

const Form = () => {
  const { url, method, body, auth, username, password, token } = useGlobalState();
  const dispatch = useDispatch();

  const handleChange = ({ target }) => {
    dispatch({ type: target.name, payload: target.value });
  };

  const handleSubmit = () => {
    const base64 = require('base-64');
    let headers;
    event.preventDefault();
    if(auth === 'basic') headers = `Basic ${base64.encode(`${username}:${password}`)}`;
    if(auth === 'bearer') headers = `Bearer ${token}`;
    fetchResponse(url, method, body, headers)
      .then(res => {
        dispatch({ type: 'SET_RES', payload: res });
        if(res.ok) {
          dispatch({ type: 'ADD_REQUESTS', payload: { url: url, method: method, body: body } });
        }
      });
  };

  const authType = () => {
    if(auth === 'basic')
      return (
        <>
          <input type="text" name="SET_USERNAME" value={username} onChange={handleChange} placeholder="username"></input>
          <input type="text" name="SET_PASSWORD" value={password} onChange={handleChange} placeholder="password"></input>
        </>
      );
    if(auth === 'bearer')
      return (
        <>
          <input type="text" name="SET_TOKEN" value={token} onChange={handleChange} placeholder="token"></input>
        </>
      );
  };

  return (
    <section className={styles.Form}>
      <form onSubmit={handleSubmit}>
        <div className={styles.Left}>
          <input type="text" name="SET_URL" value={url} onChange={handleChange} placeholder="Enter URL"></input>
          <div>
            <label><input type="radio" name="SET_METHOD" value="GET" onChange={handleChange} checked={method === 'GET'}/>GET</label>
            <label><input type="radio" name="SET_METHOD" value="POST" onChange={handleChange} checked={method === 'POST'}/>POST</label>
            <label><input type="radio" name="SET_METHOD" value="PUT" onChange={handleChange} checked={method === 'PUT'}/>PUT</label>
            <label><input type="radio" name="SET_METHOD" value="PATCH" onChange={handleChange} checked={method === 'PATCH'}/>PATCH</label>
            <label><input type="radio" name="SET_METHOD" value="DELETE" onChange={handleChange} checked={method === 'DELETE'}/>DELETE</label>
          </div>
          <textarea name="SET_BODY" value={body} onChange={handleChange} placeholder="Enter JSON Body"></textarea>
          <button type="submit">Submit</button>
        </div>
        <div className={styles.Right}>
          <div className={styles.AuthTop}>
            <h3>Authorization</h3>
            <select id="auth" name="SET_AUTH" onChange={handleChange}>
              <option name="SET_AUTH" value="none">None</option>
              <option name="SET_AUTH" value="basic">Basic Auth</option>
              <option name="SET_AUTH" value="bearer">Bearer Token</option>
            </select>
          </div>
          <div className={styles.AuthBot}>
            {authType()}
          </div>
        </div>
      </form>   
    </section>
  );
}; 

export default Form;
