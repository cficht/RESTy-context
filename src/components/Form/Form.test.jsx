import React from 'react';
import { shallow } from 'enzyme';
import Form from './Form';
import { ResponseProvider } from '../../hooks/ResponseProvider';

describe('Form component', () => {
  it('renders Form', () => {
    const wrapper = shallow(
      <ResponseProvider>
        <Form />
      </ResponseProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
