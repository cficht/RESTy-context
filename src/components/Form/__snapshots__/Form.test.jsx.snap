// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component renders Form 1`] = `
<ContextProvider
  value={
    Object {
      "dispatch": [Function],
      "state": Object {
        "body": "",
        "method": "",
        "requests": Array [],
        "res": Object {},
        "url": "",
      },
    }
  }
>
  <Form />
</ContextProvider>
`;
